src/Swarm/Docs/Pedagogy.hs

```diff
@@ -0,0 +1,201 @@
+import Swarm.Game.Scenario.Objective (objectiveGoal)
+import Swarm.Game.ScenarioInfo (ScenarioCollection, ScenarioInfoPair, flatten, loadScenariosWithWarnings, scenarioCollectionToList, scenarioPath)
+import Swarm.Language.Module (Module (..))
+import Swarm.Language.Pipeline (ProcessedTerm (..))
+import Swarm.Language.Syntax
+import Swarm.Language.Types (Polytype)
+import Swarm.TUI.Controller (getTutorials)
+import Swarm.Util (commaList)
+
+-- * Constants
+
+wikiPrefix :: Text
+wikiPrefix = "https://github.com/swarm-game/swarm/wiki/"
+
+commandsWikiPrefix :: Text
+commandsWikiPrefix = wikiPrefix <> "Commands-Cheat-Sheet#"
+
+-- * Types
+
+-- | Tutorials augmented by the set of
+-- commands that they introduce.
+-- Generated by folding over all of the
+-- tutorials in sequence.
+data CoverageInfo = CoverageInfo
+  { tutInfo :: TutorialInfo
+  , novelSolutionCommands :: Set Const
+  }
+
+-- | Tutorial scenarios with the set of commands
+-- introduced in their solution and descriptions
+-- having been extracted
+data TutorialInfo = TutorialInfo
+  { scenarioPair :: ScenarioInfoPair
+  , solutionCommands :: Set Const
+  , descriptionCommands :: Set Const
+  }
+
+-- | A private type used by the fold
+data CommandAccum = CommandAccum
+  { _encounteredCmds :: Set Const
+  , tuts :: [CoverageInfo]
+  }
+
+-- * Functions
+
+-- | Extract commands from both goal descriptions and solution code.
+extractCommandUsages :: ScenarioInfoPair -> TutorialInfo
+extractCommandUsages siPair@(s, _si) =
+  TutorialInfo siPair solnCommands $ getDescCommands s
+ where
+  solnCommands = S.fromList $ maybe [] getCommands maybeSoln
+  maybeSoln = view scenarioSolution s
+
+-- | Obtain the set of all commands mentioned by
+-- name in the tutorial's goal descriptions.
+--
+-- NOTE: It may be more robust to require that a command reference
+-- be surrounded by backticks and parse for that accordingly.
+getDescCommands :: Scenario -> Set Const
+getDescCommands s =
+  S.fromList $ mapMaybe (`M.lookup` txtLookups) allWords
+ where
+  goalTextParagraphs = concatMap (view objectiveGoal) $ view scenarioObjectives s
+  allWords = concatMap (wordsBy (not . isLetter) . T.unpack . T.toLower) goalTextParagraphs
+
+  commandConsts = filter isCmd allConst
+  txtLookups = M.fromList $ map (T.unpack . syntax . constInfo &&& id) commandConsts
+
+-- | Extract the command names from the source code of the solution.
+--
+-- NOTE: The processed solution stored in the scenario has been "decorated" in a sense;
+-- e.g. `noop` gets inserted for an empty `build {}` command.
+-- So we explicitly ignore `noop`.
+--
+-- Also, the code from `run` is not parsed transitively yet.
+getCommands :: ProcessedTerm -> [Const]
+getCommands (ProcessedTerm (Module stx _) _ _) =
+  mapMaybe isCommand nodelist
+ where
+  ignoredCommands = S.fromList [Run, Noop]
+
+  nodelist :: [Syntax' Polytype]
+  nodelist = universe stx
```

Comment:
I think this may not actually be sufficient to find all of the `Const` instances for the commands?  It might not be finding the commands hidden inside a `build` command.  Will I have to use an approach like the [`requirements'` function](https://github.com/swarm-game/swarm/blob/9e2f8b29613223b7eaded5690257abbc6553ac93/src/Swarm/Language/Requirement.hs#L177) instead of this simple `universe` function?

Reply:
There's no reason commands hidden inside `build` would not be found.  `universe` works in terms of the `Plated` instance which is derived entirely generically in terms of the auto-derived `Data` instance.  `build` is just an application of the `Build` constant to an argument; to `universe`, the `build` command is not special in any way.

Reply:
Perhaps the reason #1086 is not caught is that `give` technically shows up in the `Debug` tutorial (although it is not really explained there, you are just told to do it).  Also, the English word "give" (but NOT the command) shows up in the description of the `Type errors` tutorial.

Reply:
I realize now that the "Define" tutorial (`def.yaml`) comes *after* the "Debug" tutorial (`crash.yaml`), which is why the new tests caught the problem in `crash.yaml` but not `def.yaml`.  So the `universe` function works fine.